Add `BasicAssetLibrary` example manager

foundrytom opened the pull request:
As described in #215, this PR adds the `BasicAssetLibrary` manager to the example resources. In addition, it extends the coverage of `openassetio.test.manager.apiComplicanceSuite` to include the methods it currently supports, and adds a few quality-of-life improvements to the harness itself.

The Basic Asset Library is a very simple "map some string to another string" manager. It is incredibly basic. The intent is that over time, this can be expanded to cover basic registration with versioning and related references.

BAL deliberately includes unicode characters in its test data sets to make sure we are exercising that aspect of the API code paths on a regular basis. It also demonstrates the re-use of the test harness with a custom manager-specific suite to check its proprietary business logic.

This PR will be split up into smaller ones, but it keeps BAL along side the test harness code for now for TDD/see-it-working reasons.

## WIP

This is a WIP PR, there are still some things to be done:

- UTF-8 headers?
- Documentation around BAL is incomplete.
- Some commits need more detail in the messages.
- Project documentation needs updating to match the checks made in the `apiComplicanceSuite`
- The BAL implementation may need some Python 3 checking, a certain person hasn't really written any proper Python code since we agreed on 3 only!

It is here to solicit feedback on the following topics:

 - How do we feel about the BAL code (ignoring the incomplete docs), which is (hopefully) using PEP-8 and [black](https://pypi.org/project/black/). This follows our practice of writing example code as full PEP-8, without the naming exceptions we apply to code within the  `openassetio` package - to match our `C++` convention.

 - BAL is deliberately poorly engineered. People have a habit of using sample code in production. By attempting to make it "just about robust enough to function", this may discourage its real-world use. Consequently, there is no unit testing etc. Good idea? Bad Idea? Is this "bad enough to disuade, good enough to be useful for testing"?

 - The extensions to `apiComplianceSuite` aim to follow the "test only that the API contract is met, not any manager-specific business logic" mantra. In practice, this is actually harder to do than it sounds. I opted to go for the "disable checks unless fixtures are provided" approach, over using of `managementPolicy`.  It may seem like we should be querying that, but what happens is that you end up needing a parameterised fixture that would return a managed spec, to know you could say, resolve, which then means you can ask for _another_ parameterised fixture for some valid ref (as the spec itself isn't involved in resolve). More broadly speaking, we opted to make the official API policy that `isEntityReference` should _not_ require management policy to be queried first. So, for a manager `isEntityReference` and `resolveEntityReference` are required methods, and could be called at any time with any valid string/reference. There are many use cases where the host will not know the spec of some given reference (eg: OTIO resolve, some arbitrary resolving python script). It would be overlay chatty if the host was force to find this out for every reference it needed to resolve, and call managementPolicy for each unique spec. Certainly not practical when resolving millions of references for scene expansion. We need to make this clear in the documentation (see todo above). Are we happy with the skip-if-no-fixture approach?

foundrytom commented on resources/examples/manager/BasicAssetLibrary/tests/fixtures.py:
```diff
@@ -0,0 +1,105 @@
+    },
+    "shared": {
+        "a_valid_entity_reference": valid_ref,
+        "a_non_entity_reference": non_ref,
```
I struggled with this one. I had thought of invalid too, but wondered if this could be construed as "an entity reference that is not valid" ie. a correctly formatted string, that does not convey valid data. Rather than "a string that is definitely not formatted as an entity reference". How about:

```
"an_entity_reference"
"not_an_entity_reference"
```

feltech replied:
Hmm, good point. My concern with `a_non_entity_reference` was that, at a glance, I read it to mean "a valid reference to something that is not an entity". How about a `a_malformed_entity_reference`?

I think `not_an_entity_reference` could still fall into the trap of assuming a valid entity reference to an invalid entity, but is still better than `a_non_entity_reference`.  

I kinda like that we could establish a convention of `a_<happy path fixture>` side-by-side with `not_a_<happy path fixture>`, but I suspect, like this case, it would usually be too ambiguous.

feltech replied:
Does dropping the word "entity" help disambiguate? `a_valid_reference` / `a_malformed_reference`?

foundrytom commented on resources/examples/manager/BasicAssetLibrary/tests/fixtures.py:
```diff
@@ -0,0 +1,105 @@
+    "Test_setSettings": {
+        "shared": {
+            "some_valid_settings": {"library_path": "/dev/null"},
+            "some_settings_with_unknown_keys": {"cat": True}
```
I wondered where on the fence that sat too. We say in the API contract (or will do, if we don't already) that unknown keys should cause an error.  So I guess if we test value types, we need to state the behaviour then (so that it is in API Contract territory rather than local business logic)?

feltech replied:
Maybe the language should be "invalid entries" rather than "invalid keys"?

In the docs `setSettings` is currently blank for `Manager` and says "Todo: Document settings mechanism" for `ManagerInterface`.

foundrytom commented on python/openassetio/test/manager/apiComplianceSuite.py:
```diff
@@ -129,3 +176,113 @@ def __assertPolicyResults(self, numSpecifications, context):
+    """
+
+    def setUp(self):
+        self._a_valid_reference = self._fixtures.get('a_valid_entity_reference')
```
Great idea.

I've sketched something out [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/4a144487ed52af7119ef2e049dd751be659b1e07). I updated the behaviour a little from your design - as when I put it into practice, it felt like the default of a 'require' method to skip, was perhaps backwards. See what you think.  I also added some conveniences to avoid boilerplate in common use patterns that came up whilst prototyping adoption [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/ac77b93ef02f7d32aba3c63a677d4e573fd5f02c), [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/5bf3109c75d97cfe9d963856f23e866e81a653cc) and [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/82459fdbdd6965ab94703c1ff7744f8cb5657e72).

feltech replied:
Just occurred to me that `collectRequiredFixture` means that in practice all fixture keys must be valid variable names, which is a restriction we didn't previously have.  I guess I'm happy with that if you are?

foundrytom replied:
Cool, yeah, felt reasonable. I added some notes [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/b3f57a2e0bbdc871b5dd3022648efc31b1752eb2).

foundrytom commented on python/openassetio/test/manager/__main__.py:
```diff
@@ -54,7 +54,10 @@
 
         fixtures = {
             "identifier" : "<identifier of target plugin>",
+            "settings": { <manager_setting>: <value? },
```
🦅 👁️ , thanks! fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/dbf78cb88cda129782b1250f52e608c21301fcf8).

foundrytom commented on tests/openassetio/test/manager/resources/suite_executeSuiteTests.py:
```diff
@@ -59,11 +65,55 @@ def test_when_called_then_manager_is_the_expected_manager(self):
+    def test_fixtures_contain_no_additional_keys(self):
+        self.assertEqual(
+            self._fixtures,
+            self._fixtures,
```
Fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/543e9977c7a61b113ee559c2ccd11e03693653f9).

foundrytom commented on python/openassetio/test/manager/apiComplianceSuite.py:
```diff
@@ -82,6 +84,51 @@ def test_is_correct_type(self):
+        updated_settings = self._fixtures.get("updated_settings_with_all_keys")
+        if updated_settings is None:
+            self.skipTest("No 'updated_settings_with_all_keys' fixture provided")
+            return
```
These should all be gone, now all cases have been updated to the new fixture retrieval conveniences.

foundrytom commented on python/openassetio/test/manager/harness.py:
```diff
@@ -210,3 +237,61 @@ def assertValuesOfType(self, valueList, valueType, allowNone=False):
+            action = self.skipTest if skipTestIfMissing else self.fail
+            action(message)
+            # We should never get here as action should raise an exception
+            # either way to abort/fail the test
```
Good spot, thanks! Fixed [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/ad340e9f21e91834396ce5351e42e2ee286df76e)

foundrytom commented:
@feltech If you have a sec, I've (hopefully) addressed the issues other than the names of the fixtures. If you're happy with those other changes, I can squash them all together - which will make renaming the fixtures a bit easier.

foundrytom commented on resources/examples/manager/BasicAssetLibrary/tests/fixtures.py:
```diff
@@ -0,0 +1,105 @@
+    "Test_entityExists": {
+        "shared": {
+            "an_existing_entity_reference": an_existing_ref,
+            "a_non_existant_entity_reference": valid_ref
```
Updated [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/3813511bac4e049296bd222b75cff9c097503486).

foundrytom commented:
@feltech see [this commit](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/239/commits/431fe4f097dac31cf7f21fedf06b59d71a426cfb) and onwards for some improvements to BAL.

feltech commented on resources/examples/manager/BasicAssetLibrary/README.md:
```diff
@@ -0,0 +1,40 @@
+system.
+
+It serves to provide a minimum level of functionality to allow simple,
+repeatable demonstrations and end-to-end tests to be realized with a
```
[fix]
```suggestion
repeatable demonstrations and end-to-end tests to be realized with as
```

feltech commented on resources/examples/manager/BasicAssetLibrary/plugin/BasicAssetLibrary/bal.py:
```diff
@@ -0,0 +1,95 @@
+def make_default_settings() -> dict:
+    """
+    Generates a default settings dict for BAL.
+    Note: as a library is required. The default settings are not enough
```
[fix]
```suggestion
    Note: as a library is required, the default settings are not enough
```

feltech commented on resources/examples/manager/BasicAssetLibrary/schema.json:
```diff
@@ -0,0 +1,41 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
```
[question] Neat that we're adding a JSON schema, but curious what prompted this? Do we expect BAL to be used externally?  Or is it just that it seems good practice for completeness?

foundrytom replied:
We will use it in `otio-openassetio` for e2e tests, and any host implementation may wish to use it for simple, repeatable test-harness workflows. As such, it felt like it would be better than writing loose documentation.

feltech reviewed (approved):
Looks good, squash and merge when ready!
